implement bitfield distribution subsystem (#1368)
paritytech/polkadot: roadmap/implementers-guide/src/node/availability/bitfield-distribution.md

`ProtocolId`: `b"bitd"`

Input:
[`BitfieldDistributionMessage`](../../types/overseer-protocol.md#bitfield-distribution-message) which are gossiped to all peers, no matter if validator or not.

Output:

> - `NetworkBridge::RegisterEventProducer(ProtocolId)` in order to register ourselfs as an event provide for the protocol.

Reviewer: ```suggestion
- `NetworkBridge::RegisterEventProducer(ProtocolId)` in order to register ourself as an event provider for the protocol.
```

- `NetworkBridge::SendMessage([PeerId], ProtocolId, Bytes)`

> - `NetworkBridge::ReportPeer(PeerId, cost_or_benefit)` improve or penalize the reputiation of peer based on the message we received relative to our state.

Reviewer: ```suggestion
- `NetworkBridge::ReportPeer(PeerId, cost_or_benefit)` improve or penalize the reputation of peer based on the message we received relative to our state.
```

> - `BitfieldDistributionMessage::DistributeBitfield(relay_parent, SignedAvailabilityBitfield)` gossip a verified incoming bitfield on to interested subsystems within this validator node.

Reviewer: Is this saying that the `BitfieldDistribution` subsystem will output its own message?

Author: An initial misconception.

## Functionality

This is implemented as a gossip system. Register a [network bridge](../utility/network-bridge.md) event producer on startup.

> It is necessary to track peer connection, view change, and disconnection events, in order to maintain an index of which peers are interested in which relay parent bitfields.

Reviewer 2: This section will be rendered as a single paragraph; it feels like it might be intended to be rendered as an unordered list.

> Before gossiping incoming bitfields on, they must be checked to be signed by one of the validators

Reviewer 2: ```suggestion
Before gossiping incoming bitfields, they must be checked to be signed by one of the validators
```

"gossiping X on" feels like an odd phrase to me, very German (_eingossipen_?). If the intent is specifically about relaying incoming gossip messages, you might say "before forwarding incoming gossiped bitfields, ...". Otherwise, we can just eliminate the "on".
